pyrevolve/crevolve.pyx

```diff
@@ -0,0 +1,143 @@
+cimport revolve_c
+
+from enum import Enum
+import warnings
+
+class RevolveError(Exception):
+    pass
+    # TODO: the hardcoded limits really should be removed in a future version. This should be as easy as replacing the arrays in the C++ code with an std::vector.
+    errCodes = {
+        10: "Number of checkpoints stored exceeds hardcoded limit. Modify 'checkup' and recompile pyrevolve.",
+        11: "Number of checkpoints stored exceeds snaps. This is a bug in pyrevolve, please report it.",
+        12: "Internal error in numforw, please report a pyrevolve bug.",
+        13: "Enhancement of 'fine', increase 'snaps'. Please rerun and allow for more checkpoints to be stored.",
+        14: "Number of snaps exceeds hardcoded limit. Modify 'snapsup' and recompile pyrevolve.",
+        15: "Number of reps exceeds hardcoded limit. Modify 'repsup' and recompile pyrevolve."
+    }
+
+class Action(Enum):
+    advance   = 1
+    takeshot  = 2
+    restore   = 3
+    firstrun  = 4
+    youturn   = 5
+    terminate = 6
+    error     = 7
+
+def adjust(timesteps):
+    cdef int c_st = timesteps
+    return revolve_c.revolve_adjust(c_st)
+
+def maxrange(snapshots, timefactor):
+    cdef int c_ss = snapshots
+    cdef int c_tt = timefactor
+    return revolve_c.revolve_maxrange(c_ss, c_tt)
+
+def numforw(timesteps, snapshots):
+    cdef int c_st = timesteps
+    cdef int c_sn = snapshots
+    return revolve_c.revolve_numforw(c_st, c_sn)
+
+def expense(timesteps, snapshots):
+    cdef int c_st = timesteps
+    cdef int c_sn = snapshots
+    return revolve_c.revolve_expense(c_st, c_sn)
+
+cdef class CRevolve(object):
+    cdef revolve_c.CRevolve __r
```

Comment:
This will effectively force a singleton object `__r` shared between every `CRevolve` instance. Are you sure you don't want to define this in `__init__` as `cef revolve_c.CReveolve self.__r`?